shopinvader_api_cart/schemas/cart.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from datetime import datetime
+from typing import List
+
+from extendable_pydantic import ExtendableModelMeta
+from pydantic import BaseModel
+
+from odoo.addons.pydantic import utils
+from odoo.addons.shopinvader_schema_address.schemas import (
+    BillingAddress,
+    ShippingAddress,
+)
+
+from .amount import SaleAmount
+from .sale_order_line import SaleOrderLine
+
+
+class CartTransaction(BaseModel, metaclass=ExtendableModelMeta):
+    uuid: str | None
+    qty: float
+    product_id: int
+
+    class Config:
+        orm_mode = True
+        getter_dict = utils.GenericOdooGetter
+
+
+class CartSyncInput(BaseModel, metaclass=ExtendableModelMeta):
+    transactions: List[CartTransaction]
+
+
+class CartResponse(BaseModel, metaclass=ExtendableModelMeta):
+    uuid: str | None
+    id: int
+    state: str
+    name: str
+    date_order: datetime
+    lines: List[SaleOrderLine]
+    amount: SaleAmount | None
+    delivery: ShippingAddress | None
+    invoicing: BillingAddress | None
+    note: str | None
+
+    class Config:
+        orm_mode = True
+        getter_dict = utils.GenericOdooGetter
+
+    @classmethod
+    def from_cart(cls, odoo_rec):
+        res = cls.construct()
+        res.uuid = odoo_rec.uuid or None
+        res.id = odoo_rec.id
+        res.state = odoo_rec.state
+        res.name = odoo_rec.name
+        res.date_order = odoo_rec.date_order
+        res.lines = [
+            SaleOrderLine.from_sale_order_line(line) for line in odoo_rec.order_line
+        ]
+        res.amount = SaleAmount.from_sale_order(odoo_rec)
+        res.delivery = (
+            ShippingAddress.from_res_partner(odoo_rec.partner_shipping_id)
+            if odoo_rec.partner_shipping_id
+            else None
+        )
+        res.invoicing = (
+            BillingAddress.from_res_partner(odoo_rec.partner_invoice_id)
+            if odoo_rec.partner_invoice_id
+            else None
+        )
+        res.note = odoo_rec.note or None
```

Comment:
`construct` is [deprecated](https://docs.pydantic.dev/dev-v2/migration/). `cls(uuid=..., id=..., ...)` is better; I think.

Reply:
Okay, but it doesn't work (up to now) for SaleAmount, because this class is inherited in SaleLineAmount where another field is required. @lmignon we should have a look together at it.

Reply:
Ok to take a look together tomorrow...

Reply:
2 options here: 

* `construct` is deprecated but replaced with `model_construct`. This method creates a new instance of the class (setting default values) but does not validate data. It allows to gradually instantiate the model, when data is provided
* Using the constructor `cls(...)` validates the data, which means that in the base model, all required fields must be set. Hence `from_sale_order_line` method must take `kwargs` as params to give them in the base constructor. 

Do you have an opinion about one of these options?

Reply:
I prefer `model_construct`, then.

Reply:
IMO, `model_construct` allows a smoother development by avoiding the need to collect all the attributes required to instantiate a model before calling the constructor. If I'm not wrong, in every cases, the validation is done against the defined constrains by fastapi when building the response. @marielejeune It's something we could check...